tracker/gioui/patch/button.go

```diff
@@ -0,0 +1,196 @@
+	// Start is when the press began.
+	Start time.Time
+	// End is when the press was ended by a release or cancel.
+	// A zero End means it hasn't ended yet.
+	End time.Time
+	// Cancelled is true for cancelled presses.
+	Cancelled bool
+}
+
+// Click executes a simple programmatic click.
+func (b *Clickable) Click() {
+	b.requestClicks++
+}
+
+// Clicked calls Update and reports whether a click was registered.
+func (b *Clickable) Clicked(gtx layout.Context) bool {
+	return b.clicked(b, gtx)
+}
+
+func (b *Clickable) clicked(t event.Tag, gtx layout.Context) bool {
+	_, clicked := b.update(t, gtx)
+	return clicked
+}
+
+// Hovered reports whether a pointer is over the element.
+func (b *Clickable) Hovered() bool {
+	return b.click.Hovered()
+}
+
+// Pressed reports whether a pointer is pressing the element.
+func (b *Clickable) Pressed() bool {
+	return b.click.Pressed()
+}
+
+// History is the past pointer presses useful for drawing markers.
+// History is retained for a short duration (about a second).
+func (b *Clickable) History() []Press {
+	return b.history
+}
+
+// Layout and update the button state.
+func (b *Clickable) Layout(gtx layout.Context, w layout.Widget) layout.Dimensions {
+	return b.layout(b, gtx, w)
+}
+
+func (b *Clickable) layout(t event.Tag, gtx layout.Context, w layout.Widget) layout.Dimensions {
+	for {
+		_, ok := b.update(t, gtx)
+		if !ok {
+			break
+		}
+	}
+	m := op.Record(gtx.Ops)
+	dims := w(gtx)
+	c := m.Stop()
+	defer clip.Rect(image.Rectangle{Max: dims.Size}).Push(gtx.Ops).Pop()
+	semantic.EnabledOp(gtx.Enabled()).Add(gtx.Ops)
+	b.click.Add(gtx.Ops)
+	event.Op(gtx.Ops, t)
+	c.Add(gtx.Ops)
+	return dims
+}
+
+// Update the button state by processing events, and return the next
+// click, if any.
+func (b *Clickable) Update(gtx layout.Context) (Click, bool) {
+	return b.update(b, gtx)
+}
+
+func (b *Clickable) update(t event.Tag, gtx layout.Context) (Click, bool) {
+	for len(b.history) > 0 {
+		c := b.history[0]
+		if c.End.IsZero() || gtx.Now.Sub(c.End) < 1*time.Second {
+			break
+		}
+		n := copy(b.history, b.history[1:])
+		b.history = b.history[:n]
+	}
+	if c := b.requestClicks; c > 0 {
+		b.requestClicks = 0
+		return Click{
+			NumClicks: c,
+		}, true
+	}
+	for {
+		e, ok := b.click.Update(gtx.Source)
+		if !ok {
+			break
+		}
+		switch e.Kind {
+		case gesture.KindClick:
+			if l := len(b.history); l > 0 {
+				b.history[l-1].End = gtx.Now
+			}
+			return Click{
+				Modifiers: e.Modifiers,
+				NumClicks: e.NumClicks,
+			}, true
+		case gesture.KindCancel:
+			for i := range b.history {
+				b.history[i].Cancelled = true
+				if b.history[i].End.IsZero() {
+					b.history[i].End = gtx.Now
+				}
+			}
+		case gesture.KindPress:
+			if e.Source == pointer.Mouse {
+				gtx.Execute(key.FocusCmd{Tag: t})
+			}
+			b.history = append(b.history, Press{
+				Position: e.Position,
+				Start:    gtx.Now,
+			})
+		}
+	}
+	// qm210: removed the additional Filter for key.NameSpace
+	for {
```

Comment:
This entire for loop can be removed as we won't be getting focus and thus won't be getting any key presses either.